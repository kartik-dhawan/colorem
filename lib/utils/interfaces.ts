export interface FinalPaletteType {
  paletteGuid: string
  name: string
  hex: string[]
  rgb: number[][]
  model: string
  likes?: number
}

export type RandomColormindModelType = () => string

export interface GradientBody {
  name: string
  colors: string[]
  filter: string[]
}
