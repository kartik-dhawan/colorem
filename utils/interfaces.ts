export interface ContentfulType {
  [key: string]: any // eslint-disable-line
}

export interface EssentialItemsType {
  id: number
  body: string
}

export interface ChildrenType {
  children: JSX.Element
}

export interface SideNavItemsType {
  body: string
  id: number
  link: string
  title: string
}

export interface FooterContentObject {
  source: string
  sourceName: string
  title: string
  type: string
}

export interface PaletteDataType {
  hex: string[]
  model: string
  name: string
  paletteGuid: string
  rgb: number[][]
  __v: number
  _id: string
  likes: number
}

export interface GradientDataType {
  gradientGuid: string
  colors: string[]
  filter: string[]
  __v: number
  _id: string
  name: string
  likes: number
}

export type JSONForPaletteFunction = (
  allPalettes: PaletteDataType[], // eslint-disable-line
  count: number // eslint-disable-line
) => PaletteJSONType

export type GetLightOrDarkTextColor = (hexcode: any) => "444" | "d9d9d9" | "fff" // eslint-disable-line

export type GetContrastingColor = (hexcode: any) => string // eslint-disable-line

export type CountHandlerType = () => void

export interface PaletteColorJSONType {
  TITLE: string
  BACKGROUND: string
  TEXT: string
}

export interface PaletteJSONType {
  name?: string
  palette?: PaletteColorJSONType[]
}

export interface BoxTypeStylesType {
  [key: string]: {
    gridColumn?: {
      [key: string]: string
    }
    gridRow?: {
      [key: string]: string
    }
  }
}

export interface ColorListType {
  cid: number
  colorType: string
  colorCode: string
}

export interface InstructorObject {
  icon?: string
  instruction: string
  id: number
}

export interface AboutNavItem {
  id: number
  title: string
  route: string
}
