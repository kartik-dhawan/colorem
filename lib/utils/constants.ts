export const URLS = {
  COLORMIND_API: "http://colormind.io/api/",
}
