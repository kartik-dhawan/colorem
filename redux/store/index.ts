import { configureStore } from "@reduxjs/toolkit"
import countSlice from "../slices/countSlice"

const store = configureStore({
  reducer: {
    countSlice,
  },
})

export default store
