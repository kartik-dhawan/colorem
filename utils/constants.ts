export const LOCAL_URL = "http://localhost:3000"

export const externalLinks = {
  GITHUB: "https://github.com/kartik-dhawan",
  INSTAGRAM: "https://instagram.com/notkartikk",
  LINKEDIN: "https://www.linkedin.com/in/kartik-dhawan-13585b211/",
}
