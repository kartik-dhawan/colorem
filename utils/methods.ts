import { GetContrastingColor, GetLightOrDarkTextColor } from "./interfaces"

// lightens or darkens a color by a certain amount
export const LightenDarkenColor = (col: string, amt: number) => {
  let usePound = false

  if (col[0] == "#") {
    col = col.slice(1)
    usePound = true
  }

  const num = parseInt(col, 16)

  let r = (num >> 16) + amt

  if (r > 255) r = 255
  else if (r < 0) r = 0

  let b = ((num >> 8) & 0x00ff) + amt

  if (b > 255) b = 255
  else if (b < 0) b = 0

  let g = (num & 0x0000ff) + amt

  if (g > 255) g = 255
  else if (g < 0) g = 0

  return (usePound ? "#" : "") + (g | (b << 8) | (r << 16)).toString(16)
}

// finds out if a color is light or dark
export const lightOrDark = (color: any) => {
  // eslint-disable-line
  // letiables for red, green, blue values
  let r, g, b

  // Check the format of the color, HEX or RGB?
  if (color.match(/^rgb/)) {
    // If RGB --> store the red, green, blue values in separate letiables
    color = color.match(
      /^rgba?\((\d+),\s*(\d+),\s*(\d+)(?:,\s*(\d+(?:\.\d+)?))?\)$/
    )

    r = color[1]
    g = color[2]
    b = color[3]
  } else {
    // If hex --> Convert it to RGB: http://gist.github.com/983661
    color = +("0x" + color.slice(1).replace(color.length < 5 && /./g, "$&$&"))

    r = color >> 16
    g = (color >> 8) & 255
    b = color & 255
  }

  // HSP (Highly Sensitive Poo) equation from http://alienryderflex.com/hsp.html
  const hsp = Math.sqrt(0.299 * (r * r) + 0.587 * (g * g) + 0.114 * (b * b))

  // Using the HSP value, determine whether the color is light or dark
  if (hsp > 127.5) {
    return "light"
  } else {
    return "dark"
  }
}

// gets a contrasting color by 50% for any color
export const getContrastingColor: GetContrastingColor = (hexcode: any) => {
  // eslint-disable-line
  if (lightOrDark(hexcode) === "light") {
    return LightenDarkenColor(hexcode, -50)
  } else if (lightOrDark(hexcode) === "dark") {
    return LightenDarkenColor(hexcode, 50)
  } else {
    return "#fff"
  }
}

// returns #444 or #d9d9d9 color for text on the basis of lighter or darker background
export const getLightOrDarkTextColor: GetLightOrDarkTextColor = (
  hexcode: any // eslint-disable-line
) => {
  if (lightOrDark(hexcode) === "light") {
    return "444"
  } else if (lightOrDark(hexcode) === "dark") {
    return "d9d9d9"
  } else {
    return "fff"
  }
}

// copy to clipboard function
export const copyToClipboard = (data: string) => {
  navigator.clipboard.writeText(data)
}

export const copyHexCode = (hexcode: string) => {
  copyToClipboard(`#${hexcode}`)
}

export const copyPaletteJSON = (palette: any) => {
  copyToClipboard(JSON.stringify(palette))
}
