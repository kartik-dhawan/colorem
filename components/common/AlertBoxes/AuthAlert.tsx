import { Box, Button, Typography } from "@mui/material"
import { LoginErrorSuccess } from "../../../utils/interfaces"
import ErrorIcon from "@mui/icons-material/Error"
import CheckCircleIcon from "@mui/icons-material/CheckCircle"
import { useCallback, useState } from "react"
import { styles } from "./styles"
import { getErrorObjectByCode } from "../../../lib/auth/errorMessages"

interface AuthAlertProps {
  state: LoginErrorSuccess
}

const AuthAlert = ({ state }: AuthAlertProps) => {
  const aid = "authAlert"
  const [detailsToggle, setDetailsToggle] = useState<boolean>(false)

  const detailsToggleHandler = useCallback(() => {
    setDetailsToggle(!detailsToggle)
  }, [detailsToggle])

  return (
    <Box
      sx={styles.authAlertWrapper}
      className={aid + "Wrapper"}
      data-testid={aid + "Wrapper"}
    >
      <Box
        className={aid + "TitleWrapper"}
        data-testid={aid + "TitleWrapper"}
        sx={{
          ...styles.authAlertTitleWrapper,
          "& > svg": {
            color: state.status === "error" ? "#ca8282" : "#7aa965",
          },
        }}
      >
        {state.status === "error" && <ErrorIcon />}
        {state.status === "success" && <CheckCircleIcon />}
        <Typography
          className={aid + "Title"}
          data-testid={aid + "Title"}
          sx={{
            ...styles.authAlertTitle,
            color: state.status === "error" ? "#ca8282" : "#7aa965",
          }}
        >
          {state.error && getErrorObjectByCode(state.error?.code).title}
        </Typography>
        <Button
          className={aid + "DetailsToggle"}
          data-testid={aid + "DetailsToggle"}
          disableRipple
          sx={styles.authAlertDetailsButton}
          onClick={detailsToggleHandler}
        >
          {detailsToggle ? "(show less)" : "(show more)"}
        </Button>
      </Box>
      {detailsToggle && (
        <Typography
          className={aid + "DetailsBody"}
          data-testid={aid + "DetailsBody"}
          sx={styles.authAlertDetailsBody}
        >
          {state.error && getErrorObjectByCode(state.error?.code).message}
        </Typography>
      )}
    </Box>
  )
}

export default AuthAlert
