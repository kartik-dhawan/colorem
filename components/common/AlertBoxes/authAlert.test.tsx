import { Provider } from "react-redux"
import renderer from "react-test-renderer"
import "@testing-library/jest-dom"
import AuthAlert from "./AuthAlert"
import store from "../../../redux/store"
import { LoginErrorSuccess } from "../../../utils/interfaces"

const authError: LoginErrorSuccess = {
  status: "error",
  error: {
    code: "auth/wrong-password",
    name: "wrong password",
    message: "Firebase: Error (auth/wrong-password).",
    customData: {},
  },
}

describe("Auth Alert component", () => {
  it("-- should match snapshot", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <AuthAlert state={authError} />
        </Provider>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
