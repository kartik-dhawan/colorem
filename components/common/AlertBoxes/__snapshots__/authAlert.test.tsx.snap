// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Alert component -- should match snapshot 1`] = `
<div
  className="authAlertWrapper MuiBox-root css-g2niwc"
  data-testid="authAlertWrapper"
>
  <div
    className="authAlertTitleWrapper MuiBox-root css-1cjxdn5"
    data-testid="authAlertTitleWrapper"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="ErrorIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
      />
    </svg>
    <p
      className="MuiTypography-root MuiTypography-body1 authAlertTitle css-19hq4de-MuiTypography-root"
      data-testid="authAlertTitle"
    >
      Firebase: Error (auth/wrong-password).
    </p>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium authAlertDetailsToggle css-14ukkwu-MuiButtonBase-root-MuiButton-root"
      data-testid="authAlertDetailsToggle"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      (show more)
    </button>
  </div>
</div>
`;
